Documentation for columnar. (#151739)
elastic/elasticsearch: docs/reference/columnar/index.md

- Mappings are always flat, and object like fields in mappings are always auto-flattened.
- The original source is no longer stored. If source is requested at query time, then a flattened or columnar representation of the source is generated on the fly and returned. Optionally, the columnar source can also be generated at index time and stored to disk as doc values.
- New multi-value semantics: the original ordering of multiple values per field per document (e.g., in arrays) is preserved by default. Optionally, fields in mappings can be configured to only allow one value per document.
- Fields in mappings can be configured to reject documents that have no value for them.
- Metadata fields like `_routing` and `_id` are stored using doc values as well.
- An optimized doc values format is used by default, further reducing storage footprint. especially when combined with index sorting.

Together with index sorting, columnar mode brings Elasticsearch's storage footprint and columnar access in line with dedicated columnar stores while retaining its full search and aggregation capabilities.

## Columnar index modes [columnar-modes]

Two columnar index modes are available:

`columnar`

> :   A general-purpose columnar store with no use-case-specific defaults. Use this mode for bare indices and data streams that does not fit the logging paradigm.

Reviewer: ```suggestion
:   A general-purpose columnar store with no use-case-specific defaults. Use this mode for bare indices and data streams that do not fit the logging paradigm.
```

    Use this mode for log data.

Both modes are strictly columnar: they reject mapping-level runtime fields and prevent disabling `_source`.

## Enabling columnar mode [enabling-columnar-mode]

Set `mode` index setting at index creation time. The setting cannot be changed after the index is created.

```console
PUT my-index
{
  "settings": {
    "mode": "columnar"
  }
}
```

For log data, use a component template to enable the `logsdb_columnar` mode for all `logs-*-*` data streams:

```console
PUT _component_template/logs@custom
{
  "template": {
    "settings": {
      "mode": "logsdb_columnar"
    }
  }
}
```

By default, all data streams matching with `logs-*-*` use `logsdb`, but with the custom component template in place, all data streams matching with `logs-*-*` will use `logsdb_columnar`.

Or set it directly at index creation:

```console
PUT my-logs-index
{
  "settings": {
    "mode": "logsdb_columnar"
  }
}
```

### Index sorting [index-sorting]

An important configuration step when setting up columnar index mode is determining the index sort fields.

> Good index sorting fields allow Elasticsearch to store data more efficiently and allow queries to be executed more efficiently.

Reviewer: ```suggestion
Good index sorting fields allow storing data more efficiently and improving query response times.
```

Good index sorting fields are dependent on the use case and the data.

> The `logsdb_columnar` index mode, just like `logsdb` index mode, uses by default the `host.name` field in acendeing order and `@timestamp` field in descending order as the index sort fields.

Reviewer: ```suggestion
The `logsdb_columnar` index mode, just like `logsdb` index mode, uses by default the `host.name` field in ascending order and `@timestamp` field in descending order as the index sort fields.
```

> Hosts typically emit similar logs and therefor it makes sense logs from the host are adjacent on disk. This ensures that logs are compressed efficiently on disk.

Reviewer: ```suggestion
Hosts typically emit similar logs, so storing log entries from the same host sequentially on disk improves the efficiency of compression techniques such as run-length and delta encoding.
```

> The `@timestamp` field is also used as the index sort field because then within a host logs are stored from newest to oldest, which aligns with most queries.

Reviewer: ```suggestion
The `@timestamp` field is also used as the index sort field, with recent log entries appearing first and thus improving query performance for queries over recent data.
```

> The `columnar` index mode doesn't enable index sorting by default. Let's say you collect logs from different agents, then sorting by agent id is a good choice:

Reviewer: ```suggestion
The `columnar` index mode doesn't enable index sorting by default. Let's say you collect logs from different agents, then sorting by agent id and timestamp may be a good choice:
```

```console
PUT my-index
{
  "settings": {
    "mode": "columnar",
    "sort.field": [ "agent.id", "@timestamp" ],
    "sort.order": [ "asc", "desc" ]
  }
}
```

> If you query latency is more important than storage efficiency, then only sorting by `@timestamp` is a good choice:

Reviewer: ```suggestion
If query latency is more important than storage efficiency, then sorting by just `@timestamp` will likely improve query response times:
```

```console
PUT my-logs-index
{
  "settings": {
    "mode": "logsdb_columnar",
    "sort.field": [ "@timestamp" ],
    "sort.order": [ "desc" ]
  }
}
```

## Dynamic mappings [dynamic-mappings]

> With the columnar modes, fields that are unmapped will continue to be mapped via dynamic mappings:

Reviewer: ```suggestion
With the columnar modes, fields that are not referenced in static mappings get dynamically mapped as follows:
```

* Whole numbers are mapped as a long field type.
* Decimal numbers are mapped as a double type.
* Strings are mapped as a keyword field type.

> * Objects/arrays are mapped as one or more leaf fields (depending on the number of unmapped field paths). Mappings are flattened in columnar mode and that means that unmapped object will be flattened too. Each field path in the unmapped object will be mapped as a separate leaf field. No object fields will not be mapped.

Reviewer: ```suggestion
* Objects/arrays are mapped as one or more leaf fields (depending on the number of unmapped field paths). Mappings are flattened in columnar mode and that applies to unmapped objects too. Each leaf field under the unmapped object will be mapped as a separate leaf field. No object fields are added to the mappings.
```

> By default, for all unmapped fields doc values will be enabled. However unmapped fields will not indexed by default. This follows the design of storing each field once as doc values.

Reviewer: ```suggestion
Dynamically mapped fields are configured with doc values enabled and indexes disabled by default, in line with the columnar premise of storing each field once using doc values.
```

> The default unmapped field experience is provided by dynamic mappings. Dynamic mappings are controlled via dynamic mapping configuration and can be set to:

Reviewer: ```suggestion
Dynamic mapping behavior is controlled through configuration param `dynamic` that can be set to:
```

- **true** (default): Enables the dynamic mapping behaviour as is described above.

> - **false**: Doesn't automatically map unmapped fields. Meaning that unmapped fields will not be stored at all. Data in unmapped fields will be lost.

Reviewer: ```suggestion
- **false**: Unmapped fields are not mapped or stored. Data in unmapped fields are lost.
```

> - **strict**: Doesn't automatically map unmapped fields and fails write requests with documents that contain unmapped fields.

Reviewer: ```suggestion
- **strict**: Documents containing unmapped fields don't get indexed, raising indexing errors instead.
```

> - **runtime**: This is not supported in columnar mode.

Reviewer: ```suggestion
- **runtime**: This option is not supported in columnar mode.
```

## Auto flattening [auto-flattening]

Mappings are always flattened. When defining mappings, any object field mapper is removed and leaf field mappings are created for each field path.

> The same applied with dynamic mapping updates during indexing.

Reviewer: ```suggestion
The same applies to dynamic mapping updates during indexing.
```

> If `object` field mapping are flattened, the `enabled` and `dynamic` settings are preserved.

Reviewer: ```suggestion
During `object` flattening, the `enabled` and `dynamic` settings are preserved and separately tracked.
```

> For `passthrough` field mappings, the `priority` setting is preserved.

Reviewer: ```suggestion
Same applies to `passthrough` fields, along with their `priority` setting.
```

          "ip":   { "type": "ip" }
        }
      },
      "labels": {
        "type": "passthrough",
        "priority": 10,
        "properties": {
          "env": { "type": "keyword" }
        }
      }
    }
  }
}
```

The processed mapping (`GET my-index/_mapping`) shows the object mappers removed and their settings captured under `prefix_properties`:

```json
{
  "my-index": {
    "mappings": {
      "prefix_properties": {
        "attributes": {
          "dynamic": "false"
        },
        "labels": {
          "passthrough": 10
        }
      },
      "properties": {
        "attributes.host": { "type": "keyword" },
        "attributes.ip":   { "type": "ip" },
        "labels.env":      { "type": "keyword" }
      }
    }
  }
}
```

The `attributes` and `labels` object mappers are gone from `properties`; only the flat dotted-path leaf fields remain. The `dynamic: false` from `attributes` is preserved under `prefix_properties.attributes.dynamic`, preventing new fields under `attributes.*` from being auto-mapped at index time. The `priority: 10` from `labels` is preserved under `prefix_properties.labels.passthrough`.

## Limitations [columnar-limitations]

> The following features are not supported in columnar index modes:

Reviewer: Let's also call out that `dynamic:false` and `enabled:false` leads to data loss. Can't stress that enough.

- **nested field type**: The nested field type is not supported in columnar index modes.
- **Mapping-level runtime fields**: Defining runtime fields in the index mappings is rejected. Runtime fields can still be defined on individual search requests.

> - **Disabling doc values**: Mapped fields cannot disable doc values. Setting `doc_values` to `false` in mappings will be ignored.

Reviewer: ```suggestion
- **Disabling doc values**: Mapped fields cannot disable doc values. Setting `doc_values` to `false` leads to mapping errors. The only exception is multi-fields, where it's often desirable to store doc values for just one and use different index configurations for the rest.
```

> - **Field type without doc values support**: Field types that don't support doc values are not supported in columnar mode (e.g. `search_as_you_type`).

Reviewer: ```suggestion
- **Incompatible field types**: Field types that don't support doc values are not supported in columnar mode (e.g. `search_as_you_type`).
```

---
navigation_title: Columnar
applies_to:

>   stack: preview

Reviewer: ```suggestion
  stack: preview 9.5
```

* Decimal numbers are mapped as a double type.
* Strings are mapped as a keyword field type.
* Objects/arrays are mapped as one or more leaf fields (depending on the number of unmapped field paths). Mappings are flattened in columnar mode and that applies to unmapped objects too. Each leaf field under the unmapped object will be mapped as a separate leaf field. No object fields are added to the mappings.

Dynamically mapped fields are configured with doc values enabled and indexes disabled by default, in line with the columnar premise of storing each field once using doc values.

Dynamic mapping behavior is controlled through configuration param `dynamic` that can be set to:
- **true** (default): Enables the dynamic mapping behaviour as is described above.
- **false**: Unmapped fields are not mapped or stored. Data in unmapped fields are lost.
- **strict**: Documents containing unmapped fields don't get indexed, raising indexing errors instead.
- **runtime**: This option is not supported in columnar mode.

### Dynamic false [dynamic-false]

If dynamic false has been configured, then only fields that are explicitly mapped in the mappings are stored.

> Fields that are not explicitly mapped in the mappings are not stored and therefor are lost.

Reviewer: ```suggestion
Fields that are not explicitly mapped in the mappings are not stored and therefore lost.
```

    "sort.order": [ "desc" ]
  }
}
```

## Dynamic mappings [dynamic-mappings]

With the columnar modes, fields that are not referenced in static mappings get dynamically mapped as follows:
* Whole numbers are mapped as a long field type.
* Decimal numbers are mapped as a double type.
* Strings are mapped as a keyword field type.
* Objects/arrays are mapped as one or more leaf fields (depending on the number of unmapped field paths). Mappings are flattened in columnar mode and that applies to unmapped objects too. Each leaf field under the unmapped object will be mapped as a separate leaf field. No object fields are added to the mappings.

Dynamically mapped fields are configured with doc values enabled and indexes disabled by default, in line with the columnar premise of storing each field once using doc values.

Dynamic mapping behavior is controlled through configuration param `dynamic` that can be set to:
- **true** (default): Enables the dynamic mapping behaviour as is described above.
- **false**: Unmapped fields are not mapped or stored. Data in unmapped fields are lost.
- **strict**: Documents containing unmapped fields don't get indexed, raising indexing errors instead.

> - **runtime**: This option is not supported in columnar mode.

Reviewer 2: This phrasing seems awkward.

>configuration param `dynamic` that can be set to: ... **runtime**: this option is not supported

 Maybe something like

```suggestion
Note that the *runtime* option is not supported in columnar mode.
```

---
navigation_title: Columnar
applies_to:
  stack: preview 9.5
  serverless: preview
---

# Columnar index mode [columnar-index-mode]

> With columnar mode enabled, Elasticsearch becomes a full **analytical and search columnar store**. Setting `index.mode` to `columnar` or `logsdb_columnar` activates a set of changes that collectively align Elasticsearch's storage model with dedicated columnar stores:

Reviewer 3: This page is marked as applicable to Serverless, but those settings are marked as applicable to Stack only. If the method for turning on columnar mode differs across those contexts, I think we should make this intro more generic. For example:

```suggestion
When you turn on columnar mode, {{es}} becomes a full **analytical and search columnar store**.
You activate a set of changes that collectively align the {{es}} storage model with dedicated columnar stores:
```

Reviewer 4: @jordan-powers @parkertimmins can you check this and the rest of the comments by Lisa?

On this one, I guess that those setting are also applicable to serverless and should be marked as such? Correct me if I am wrong

- Fields are stored **once, as doc values only**. Non-text fields are not indexed by default, eliminating the storage cost of maintaining redundant index structures. Text fields remain indexed by default to support full-text search.

> - For non-indexed fields, [doc values skippers](/reference/elasticsearch/mapping-reference/doc-values.md#doc-values-skippers) are enabled by default. Doc values skippers are compact skip lists with metadata (for example, min and max values) to avoid scanning large blocks of data when executing a query.

Reviewer 3: I presume these are metadata field names (if not they should be expanded to the full words "maximum" and "minimum")
```suggestion
- For non-indexed fields, [doc values skippers](/reference/elasticsearch/mapping-reference/doc-values.md#doc-values-skippers) are enabled by default. Doc values skippers are compact skip lists with metadata (for example, `min` and `max` values) to avoid scanning large blocks of data when running a query.
```

Reviewer 2: I don't think the skipper metadata is user-visible, it's an internal implementation detail. So updating this to use the full "minimum" and "maximum" makes sense to me.

- Fields in mappings can be configured to reject documents that have no value for them.
- Metadata fields like `_routing` and `_id` are stored using doc values as well.
- An optimized doc values format is used by default, further reducing storage footprint, especially when combined with index sorting.

Together with index sorting, columnar mode brings Elasticsearch's storage footprint and columnar access in line with dedicated columnar stores while retaining its full search and aggregation capabilities.

## Columnar index modes [columnar-modes]

Two columnar index modes are available:

`columnar`
:   A general-purpose columnar store with no use-case-specific defaults. Use this mode for bare indices and data streams that do not fit the logging paradigm.

`logsdb_columnar`
:   A columnar store with logging-oriented defaults. It inherits all behavior of the `columnar` mode and additionally:
    - Applies a default mapping that includes a `@timestamp` field (and optionally `host.name`).
    - Enables index sorting if `@timestamp` and `host.name` mappings exist.

    Use this mode for log data.

> Both modes are strictly columnar: they reject mapping-level runtime fields and prevent disabling `_source`.

Reviewer 3: ```suggestion
Both modes are strictly columnar: they reject mapping-level runtime fields and prevent turning off `_source`.
```

> ## Enabling columnar mode [enabling-columnar-mode]

Reviewer 3: ```suggestion
## Turning on columnar mode [enabling-columnar-mode]
```

> Set `mode` index setting at index creation time. The setting cannot be changed after the index is created.

Reviewer 3: > Set `mode` index setting at index creation time. The setting cannot be changed after the index is created.

Does this apply in both Stack and Serverless contexts? Elsewhere I read that on Serverless, columnar mode is part of the project type configuration.

Reviewer 2: I'm not sure how we intend to enable columnar in serverless. But if it is configured differently than in stateful, we should definitely clarify that here.

Reviewer 4: @martijnvg for your feedback. Also for your final review of the whole doc with all the updates that we have made while you were away before we publish it

Author: > Elsewhere I read that on Serverless, columnar mode is part of the project type configuration.

@lcawl can you share where you read this?

The columnar index modes are enabled in the same way in both serverless and stack contexts.

Reviewer 3: I've added suggestions to make that clear in the `index.mode` settings. I'd also read it in an internal Google doc that "On Serverless, where users manage projects rather than indices, Columnar Mode becomes part of the project type configuration", but if that's no longer the case that's great.

```console
PUT my-index
{
  "settings": {
    "mode": "columnar"
  }
}
```

For log data, use a component template to enable the `logsdb_columnar` mode for all `logs-*-*` data streams:

```console
PUT _component_template/logs@custom
{
  "template": {
    "settings": {
      "mode": "logsdb_columnar"
    }
  }
}
```

By default, all data streams matching with `logs-*-*` use `logsdb`, but with the custom component template in place, all data streams matching with `logs-*-*` will use `logsdb_columnar`.

Or set it directly at index creation:

```console
PUT my-logs-index
{
  "settings": {
    "mode": "logsdb_columnar"
  }
}
```

### Index sorting [index-sorting]

> An important configuration step when setting up columnar index mode is determining the index sort fields.

Reviewer 3: Not mandatory, but rewriting to be less passive.
```suggestion
When you set up columnar index mode, you must determine the index sort fields.
```

> Good index sorting fields allow storing data more efficiently and improving query response times.

Reviewer 3: Not mandatory, I just don't like "allow" :)
```suggestion
Good index sorting fields facilitate more efficient data storage and improved query response times.
```

Good index sorting fields are dependent on the use case and the data.

The `logsdb_columnar` index mode, like `logsdb` index mode, uses by default the `host.name` field in ascending order and `@timestamp` field in descending order as the index sort fields.
Hosts typically emit similar logs, so storing log entries from the same host sequentially on disk improves the efficiency of compression techniques such as run-length and delta encoding.
The `@timestamp` field is also used as the index sort field with recent log entries appearing first, improving query performance for queries over recent data.

> The `columnar` index mode doesn't enable index sorting by default. Let's say you collect logs from different agents, then sorting by agent id and timestamp might be a good choice:

Reviewer 3: ```suggestion
The `columnar` index mode doesn't enable index sorting by default. If you collect logs from different agents, sorting by agent ID and timestamp might be a good choice:
```

```console
PUT my-index
{
  "settings": {
    "mode": "columnar",
    "sort.field": [ "agent.id", "@timestamp" ],
    "sort.order": [ "asc", "desc" ]
  }
}
```

> If query latency is more important than storage efficiency, then sorting by just `@timestamp` will likely improve query response times:

Reviewer 3: Not sure of the level of confidence in this assertion, so suggesting "might" instead of "will likely"
```suggestion
If query latency is more important than storage efficiency, sorting by just `@timestamp` might improve query response times:
```

```console
PUT my-logs-index
{
  "settings": {
    "mode": "logsdb_columnar",
    "sort.field": [ "@timestamp" ],
    "sort.order": [ "desc" ]
  }
}
```

## Dynamic mappings [dynamic-mappings]

With the columnar modes, fields that are not referenced in static mappings get dynamically mapped as follows:
* Whole numbers are mapped as a long field type.
* Decimal numbers are mapped as a double type.
* Strings are mapped as a keyword field type.

> * Objects/arrays are mapped as one or more leaf fields (depending on the number of unmapped field paths). Mappings are flattened in columnar mode and that applies to unmapped objects too. Each leaf field under the unmapped object will be mapped as a separate leaf field. No object fields are added to the mappings.

Reviewer 3: ```suggestion
* Objects and arrays are mapped as one or more leaf fields (depending on the number of unmapped field paths). Mappings are flattened in columnar mode and that applies to unmapped objects too. Each leaf field under the unmapped object will be mapped as a separate leaf field. No object fields are added to the mappings.
```

> Dynamically mapped fields are configured with doc values enabled and indexes disabled by default, in line with the columnar premise of storing each field once using doc values.

Reviewer 3: ```suggestion
Dynamically mapped fields are configured with doc values enabled and indexes turned off by default. This behavior aligns with the columnar premise of storing each field once using doc values.
```

> Dynamically mapped fields are configured with doc values enabled and indexes disabled by default, in line with the columnar premise of storing each field once using doc values.

Reviewer 3: ```suggestion
Dynamically mapped fields are configured with doc values turned on and indexes turned off by default, in line with the columnar premise of storing each field once using doc values.
```

> Dynamic mapping behavior is controlled through configuration param `dynamic` that can be set to:

Reviewer 3: Is this target correct? If yes, it seems like that page (https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/dynamic) also needs updates
```suggestion
Dynamic mapping behavior is controlled by the [`dynamic`](/reference/elasticsearch/mapping-reference/dynamic.md) configuration parameter, which can be set to:
```

> - **true** (default): Enables the dynamic mapping behaviour as is described above.

Reviewer 3: These little spelling and terminology nits come from https://www.elastic.co/docs/contribute-docs/style-guide/accessibility#terms and https://www.elastic.co/docs/contribute-docs/style-guide/grammar-spelling#spelling
```suggestion
- `true` (default): Turns on the dynamic mapping behavior as described in the preceding paragraphs.
```

> - **false**: Unmapped fields are not mapped or stored. Data in unmapped fields are lost.

Reviewer 3: ```suggestion
- `false`: Unmapped fields are not mapped or stored. Data in unmapped fields are lost.
```

> - **strict**: Documents containing unmapped fields don't get indexed, raising indexing errors instead.

Reviewer 3: ```suggestion
- `strict`: Documents containing unmapped fields don't get indexed, raising indexing errors instead.
```

> Note that the *runtime* option is not supported in columnar mode.

Reviewer 3: ```suggestion
Note that the `runtime` option is not supported in columnar mode.
```

> ### Dynamic false [dynamic-false]

Reviewer 3: IMO this paragraph doesn't warrant a new section. Rather, I'd remove the section title and/or turn it into an [admonition](https://docs-v3-preview.elastic.dev/elastic/docs-builder/tree/main/syntax/admonitions). For example:
```suggestion
:::{note}
```

Reviewer 2: Good point, although since this could potentially lead to data loss, I'll bump it to a `warning`.

> If dynamic false has been configured, then only fields that are explicitly mapped in the mappings are stored.

Reviewer 3: This page doesn't make it clear how to configure that option, but I think it's something like this:

> If you configure `"dynamic": false` in the index mappings, only fields that are explicitly mapped are stored.

> Fields that are not explicitly mapped in the mappings are not stored and therefore lost.

Reviewer 3: If you accept my earlier suggestion about turning this into an admonition, you need to close it:
```suggestion
Fields that are not explicitly mapped in the mappings are not stored and therefore lost.
:::
```

## Auto flattening [auto-flattening]

> Mappings are always flattened. When defining mappings, object and passthrough field mappers are removed and leaf field mappings are created for each field path.

Reviewer 3: Per your updates on other pages, this is a columnar-specific behavior so IMO we ought to make that clear in the first sentence. I'm not certain whether the second sentence is something performed by users or automatically:
```suggestion
If you use columnar mode, mappings are always flattened. When you define mappings, object and passthrough field mappers are removed and leaf field mappings are created for each field path.
```

The same applies to dynamic mapping updates during indexing.

During `object` flattening, the `enabled` and `dynamic` settings are preserved and separately tracked.
Same applies to `passthrough` fields, along with their `priority` setting.

For example, given a mapping with an `attributes` object (`dynamic: false`) and a `labels` passthrough field (`priority: 10`):

```console
PUT my-index
{
  "settings": {
    "mode": "columnar"
  },
  "mappings": {
    "properties": {
      "attributes": {
        "type": "object",
        "dynamic": false,
        "properties": {
          "host": { "type": "keyword" },
          "ip":   { "type": "ip" }
        }
      },
      "labels": {
        "type": "passthrough",
        "priority": 10,
        "properties": {
          "env": { "type": "keyword" }
        }
      }
    }
  }
}
```

> The processed mapping (`GET my-index/_mapping`) shows the object mappers removed and their settings captured under `prefix_properties`:

Reviewer 3: ```suggestion
The processed mapping shows the object mappers removed and their settings captured under `prefix_properties`. For example, `GET my-index/_mapping` returns:
```

```json
{
  "my-index": {
    "mappings": {
      "prefix_properties": {
        "attributes": {
          "dynamic": "false"
        },
        "labels": {
          "passthrough": 10
        }
      },
      "properties": {
        "attributes.host": { "type": "keyword" },
        "attributes.ip":   { "type": "ip" },
        "labels.env":      { "type": "keyword" }
      }
    }
  }
}
```

The `attributes` and `labels` object mappers are gone from `properties`; only the flat dotted-path leaf fields remain. The `dynamic: false` from `attributes` is preserved under `prefix_properties.attributes.dynamic`, preventing new fields under `attributes.*` from being auto-mapped at index time. The `priority: 10` from `labels` is preserved under `prefix_properties.labels.passthrough`.

## Columnar `_source` [columnar-source]

> Columnar index modes don't store the original JSON `_source` on disk. Two `_source` modes are supported:

Reviewer 3: > Columnar index modes don't store the original JSON `_source` on disk. Two `_source` modes are supported:

Not mandatory in this PR, but it would be nice to provide more guidance about when/why you'd pick one over the other.

**Synthetic columnar `_source`**

> :   Reconstructs a flattened representation of `_source` from doc values at query time. Requires an [Enterprise license](https://www.elastic.co/subscriptions). For more information, see [Synthetic `_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md#synthetic-source).

Reviewer 3: We typically don't mention specific licenses in the docs since the source of truth for that is in the linked page:
```suggestion
:   Reconstructs a flattened representation of `_source` from doc values at query time. Requires an [appropriate license](https://www.elastic.co/subscriptions) for synthetic `_source`. For more information, refer to [Synthetic `_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md#synthetic-source).
```

**Columnar stored `_source`**
:   Materializes and stores the columnar `_source` representation on disk at index time as doc values. Used automatically when synthetic columnar `_source` is not licensed, and can also be configured explicitly to speed up `_source` retrieval. For more information, see [Columnar source](/reference/elasticsearch/mapping-reference/mapping-source-field.md#columnar-stored).

## Limitations [columnar-limitations]

The following features are not supported in columnar index modes:

> - **nested field type**: The nested field type is supported in a limited fashion in the columnar index modes. Nesting of nested field types is not supported.

Reviewer 3: ```suggestion
- **Nested field type**: The [nested field type](/reference/elasticsearch/mapping-reference/nested.md) is supported in a limited fashion in the columnar index modes. Nesting of nested field types is not supported.
```

- **Mapping-level runtime fields**: Defining runtime fields in the index mappings is rejected. Runtime fields can still be defined on individual search requests.

> - **Disabling doc values**: Mapped fields cannot disable doc values. Setting `doc_values` to `false` leads to mapping errors. The only exception is multi-fields, where it's often desirable to store doc values for just one and use different index configurations for the rest.

Reviewer 3: ```suggestion
- **Turning off doc values**: Mapped fields cannot turn off doc values. Setting `doc_values` to `false` leads to mapping errors. The only exception is multi-fields, where it's often desirable to store doc values for just one and use different index configurations for the rest.
```

- **Incompatible field types**: Field types that don't support doc values are not supported in columnar mode (for example, `search_as_you_type`).

> - **Disabling `_source`**: Setting `"_source": {"enabled": false}` is not allowed.

Reviewer 3: ```suggestion
- **Turning off `_source`**: Setting `"_source": {"enabled": false}` is not allowed.
```

---
navigation_title: Columnar
applies_to:
  stack: preview 9.5
  serverless: preview
---

# Columnar index mode [columnar-index-mode]

With columnar mode enabled, Elasticsearch becomes a full **analytical and search columnar store**. Setting `index.mode` to `columnar` or `logsdb_columnar` activates a set of changes that collectively align Elasticsearch's storage model with dedicated columnar stores:

- Fields are stored **once, as doc values only**. Non-text fields are not indexed by default, eliminating the storage cost of maintaining redundant index structures. Text fields remain indexed by default to support full-text search.
- For non-indexed fields, [doc values skippers](/reference/elasticsearch/mapping-reference/doc-values.md#doc-values-skippers) are enabled by default. Doc values skippers are compact skip lists with metadata (for example, min and max values) to avoid scanning large blocks of data when executing a query.
- Mappings are always flat, and object and passthrough fields in mappings are always auto-flattened. Nested fields are not auto flattened.

> - The original source is no longer stored. If source is requested at query time, then a flattened or columnar representation of the source is generated on the fly and returned. Optionally, the columnar source can also be generated at index time and stored to disk as doc values.

Reviewer 4: @lcawl we have it below on the "Synthetic columnar `_source`" section (line 206), but do you think that we should also add here a mention on this requiring an appropriate license (Enterprise in this case)?

Reviewer 3: I think since it's covered in greater detail in the subsequent section, it's worth mentioning and linking for more detail. For example:
```suggestion
- Depending on your license, you can choose between two [`_source` modes](#columnar-source). If the original source is no longer stored, a flattened or columnar representation of the source is generated automatically when it's requested at query time. Alternatively, the columnar source can be generated at index time and stored to disk as doc values.
```

Reviewer 2: I mostly agree with the updates, except for `If the original source is no longer stored`. In columnar, the original source is never stored. You can either store no source (reconstructing on the fly), or a modified "columnar" source.

```suggestion
- Depending on your license, you can choose between two [`_source` modes](#columnar-source). If using synthetic source, a flattened or columnar representation of the source is generated automatically when it's requested at query time. Alternatively, this columnar source can be generated at index time and stored to the disk as doc values.
```

- New multi-value semantics: the original ordering of multiple values per field per document (for example, in arrays) is preserved by default. Optionally, fields in mappings can be configured to only allow one value per document.
- Fields in mappings can be configured to reject documents that have no value for them.
- Metadata fields like `_routing` and `_id` are stored using doc values as well.
- An optimized doc values format is used by default, further reducing storage footprint, especially when combined with index sorting.

Together with index sorting, columnar mode brings Elasticsearch's storage footprint and columnar access in line with dedicated columnar stores while retaining its full search and aggregation capabilities.

## Columnar index modes [columnar-modes]

Two columnar index modes are available:

`columnar`

> :   A general-purpose columnar store with no use-case-specific defaults. Use this mode for bare indices and data streams that do not fit the logging paradigm.

Reviewer 4: @jordan-powers @parkertimmins why do we mention data streams here? Columnar does not work on data streams, does it? Or would it just be an index that can't rollover? But what about ILM rules etc?

```suggestion
:   A general-purpose columnar store with no use-case-specific defaults. Use this mode for use cases that do not fit the logging paradigm.
```

Reviewer 2: As far as I'm aware, `columnar` is just another index mode which should work fine with plain data streams.

Reviewer 4: Yes you are right Jordan, I got confused by TSDB/LogsDB for a moment. As long as there is a timestamp there, they can you regular data streams. Resolving

          "env": { "type": "keyword" }
        }
      }
    }
  }
}
```

The processed mapping shows the object mappers removed and their settings captured under `prefix_properties`. For example, `GET my-index/_mapping` returns:

```json
{
  "my-index": {
    "mappings": {
      "prefix_properties": {
        "attributes": {
          "dynamic": "false"
        },
        "labels": {
          "passthrough": 10
        }
      },
      "properties": {
        "attributes.host": { "type": "keyword" },
        "attributes.ip":   { "type": "ip" },
        "labels.env":      { "type": "keyword" }
      }
    }
  }
}
```

The `attributes` and `labels` object mappers are gone from `properties`; only the flat dotted-path leaf fields remain. The `dynamic: false` from `attributes` is preserved under `prefix_properties.attributes.dynamic`, preventing new fields under `attributes.*` from being auto-mapped at index time. The `priority: 10` from `labels` is preserved under `prefix_properties.labels.passthrough`.

## Columnar `_source` [columnar-source]

Columnar index modes don't store the original JSON `_source` on disk. Two `_source` modes are supported:

**Synthetic columnar `_source`**

> :   Reconstructs a flattened representation of `_source` from doc values at query time. Requires an [appropriate license](https://www.elastic.co/subscriptions) for synthetic `_source`. For more information, refer to [Synthetic `_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md#synthetic-source).

Reviewer 4: Updating to align with terminology used on the subscription page

```suggestion
:   Reconstructs a flattened representation of `_source` from doc values at query time. Requires an [appropriate license](https://www.elastic.co/subscriptions) for synthetic columnar `_source`. For more information, refer to [Synthetic `_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md#synthetic-source).
```

---
navigation_title: Columnar
applies_to:
  stack: preview 9.5
  serverless: preview
---

# Columnar index mode [columnar-index-mode]

When you turn on columnar mode, {{es}} becomes a full **analytical and search columnar store**.

> You activate a set of changes that collectively align the {{es}} storage model with dedicated columnar stores:

Reviewer 3: Now that I've merged https://github.com/elastic/docs-content/pull/7594 it might be helpful to explain the different purposes of these pages:

```suggestion
For a conceptual introduction to when and why to use it, go to [Columnar index mode](docs-content://manage-data/data-store/columnar.md). This page describes how to turn it on and configure index sorting and provides more details about`_source` modes and limitations.

You activate a set of changes that collectively align the {{es}} storage model with dedicated columnar stores:
```

---
navigation_title: Columnar
applies_to:
  stack: preview
  serverless: preview
---

# Columnar index mode [columnar-index-mode]

With columnar mode enabled, Elasticsearch becomes a full **analytical and search columnar store**. Setting `index.mode` to `columnar` or `logsdb_columnar` activates a set of changes that collectively align Elasticsearch's storage model with dedicated columnar stores:

- Fields are stored **once, as doc values only**. All non-text fields are not indexed by default, eliminating the storage cost of maintaining redundant index structures.

> - Mappings are always flat, and object like fields in mappings are always auto-flattened.

Reviewer: We should call out how pass-through gets handled.

> - The original source will no longer be stored. If source is requested at query time, then a flattened or columnar representation of the source is generated on the fly and returned. Optionally, the columnar source can also be generated at index time and stored to disk as doc values.

Reviewer: ```suggestion
- The original source is no longer stored. If source is requested at query time, then a flattened or columnar representation of the source is generated on the fly and returned. Optionally, the columnar source can also be generated at index time and stored to disk as doc values.
```

- New multi-value semantics: the original ordering of multiple values per field per document (e.g., in arrays) is preserved by default. Optionally, fields in mappings can be configured to only allow one value per document.
- Fields in mappings can be configured to reject documents that have no value for them.

> - Metadata like `_routing` and `_id` are stored using doc values.

Reviewer: ```suggestion
- Metadata fields like `_routing` and `_id` are stored using doc values as well.
```

> - Optimized doc values format is used by default, further reducing storage footprint.

Reviewer: ```suggestion
- An optimized doc values format is used by default, further reducing storage footprint. especially when combined with index sorting.
```

> 

Reviewer: Do we need a bullet for `dynamic`?

Author: yes, and enable: false

These changes bring Elasticsearch's storage footprint and columnar access in line with dedicated columnar stores while retaining its full search and aggregation capabilities.

## Columnar index modes [columnar-modes]

Two columnar index modes are available:

`columnar`

> :   A general-purpose columnar store with no use-case-specific defaults. Use this mode for data that does not fit the logging model.

Reviewer: ```suggestion
:   A general-purpose columnar store with no use-case-specific defaults. Use this mode for bare indices and data streams that does not fit the logging paradigm.
```

`logsdb_columnar`

> :   A columnar store with logging-oriented defaults. It inherits all behavior of the `columnar` mode and additionally:

Reviewer: What about marking the `message` field having an index? We should probably call it out that it still gets one?

Author: I don't think this is specifically for logsdb_columnar. In any columnar mode all text fields are indexed by default. Message field is mapped as match_only_text typically and there it will have an index.

    - Applies a default mapping that includes a `@timestamp` field (and optionally `host.name`).

>     - Enables sort-field-based routing by default (`index.logsdb.route_on_sort_fields: true`), which routes documents based on the configured index sort fields rather than a routing path.

Reviewer: This may lead to hotspots, we should call this out at some point. See also https://www.elastic.co/docs/manage-data/data-store/data-streams/logs-data-stream-configure#logsdb-sort-routing

Reviewer: I actually don't see this in the logic. Remove this line for now?

Both modes are strictly columnar: they reject mapping-level runtime fields and prevent disabling `_source`.

## Enabling columnar mode [enabling-columnar-mode]

Set `index.mode` at index creation time. The setting cannot be changed after the index is created.

```console
PUT my-index
{
  "settings": {
    "index": {
      "mode": "columnar"
    }
  }
}
```

For log data, use a component template to enable the `logsdb_columnar` mode for all `logs-*-*` data streams:

```console
PUT _component_template/logs@custom
{
  "template": {
    "settings": {
      "index": {
        "mode": "logsdb_columnar"
      }
    }
  }
}
```

Or set it directly at index creation:

```console
PUT my-logs-index
{
  "settings": {
    "index": {
      "mode": "logsdb_columnar"
    }
  }
}
```

## Limitations [columnar-limitations]

The following features are not supported in columnar index modes:

- **Mapping-level runtime fields**: Defining runtime fields in the index mappings is rejected. Runtime fields can still be defined on individual search requests.
- **Disabling doc values**: Mapped fields cannot disable doc values.
- **Disabling `_source`**: Setting `"_source": {"enabled": false}` is not allowed.

> - **Stored source mode**: The traditional `stored` source mode is not supported; only `synthetic` and `columnar_stored` are available.

Reviewer: Add a cross reference to the `columnar_stored` definition, and maybe a quick explanation here.

Reviewer: Do we have an entry for columnar_stored in our docs? @jordan-powers too.

Reviewer 4: We need to also add somewhere that synthetic mode is only available for Enterprise Tier and call it "Synthetic columnar `_source`"
